### PR TITLE
Fixed #32489 -- Added iter_test_cases() to iterate over a TestSuite.

### PR DESCRIPTION
**Pull request by cjerdonek**
https://code.djangoproject.com/ticket/32489

### CODE REVIEW

File: django/test/runner.py [hunk 1]
```diff
@@ -800,19 +800,13 @@ def partition_suite_by_type(suite, classes, bins, reverse=False):
...
     """
-    suite_class = type(suite)
-    if reverse:
-        suite = reversed(tuple(suite))
-    for test in suite:
-        if isinstance(test, suite_class):
-            partition_suite_by_type(test, classes, bins, reverse=reverse)
+    for test in iter_test_cases(suite, reverse=reverse):
+        for i in range(len(classes)):
+            if isinstance(test, classes[i]):
```
Comment (jacobtylerwalls) [inline django/test/runner.py:805]: Fan of `enumerate`?
```suggestion
        for i, klass in enumerate(classes):
            if isinstance(test, klass):
```

Comment (cjerdonek) [reply django/test/runner.py:805]: I was avoiding doing any refactoring of `partition_suite_by_type()` to keep this PR focused just on `iter_test_cases()`. (I was actually planning to file a ticket on simplifying `partition_suite_by_type()` in a bunch of ways after this is merged. To give one example, you can actually dispense with the indices entirely and just use `zip()`.)

**Comment by felixxm** [comment]
@cjerdonek Thanks :+1: 

Can we add changes to other methods mentioned in the ticket? I would like to be sure that `iter_test_cases()` can be used in all of them.

**Comment by cjerdonek** [comment]
> Can we add changes to other methods mentioned in the ticket?

@felixxm Done

**Comment by felixxm** [comment]
@cjerdonek This looks great, thanks again :rocket:

**Comment by cjerdonek** [comment]
@felixxm No prob, and thanks for your time with this!

**Comment by felixxm** [comment]
Ahhh I found one issue with `partition_suite_by_case()`, try:
```
./runtests.py test_runner --settings=test_sqlite
```
it crashes.

**Comment by cjerdonek** [comment]
Can you paste the error? What does it have to do with?

**Comment by cjerdonek** [comment]
Nm, I see it.

**Comment by felixxm** [comment]
```
======================================================================
ERROR: test_autoincrement_reset2 (test_runner.tests.AutoIncrementResetTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/django/django/test/testcases.py", line 274, in _setup_and_call
    self._pre_setup()
  File "/django/django/test/testcases.py", line 960, in _pre_setup
    self._fixture_setup()
  File "/django/django/test/testcases.py", line 1002, in _fixture_setup
    self._reset_sequences(db_name)
  File "/django/django/test/testcases.py", line 991, in _reset_sequences 
    no_style(), conn.introspection.sequence_list())
  File "/django/django/db/backends/base/introspection.py", line 130, in sequence_list 
    sequence_list.extend(self.get_sequences(cursor, model._meta.db_table, model._meta.local_fields))
  File "/django/django/db/backends/sqlite3/introspection.py", line 112, in get_sequences 
    pk_col = self.get_primary_key_column(cursor, table_name)
  File "/django/django/db/backends/sqlite3/introspection.py", line 212, in get_primary_key_column
    raise ValueError("Table %s does not exist" % table_name)
ValueError: Table test_runner_person does not exist

======================================================================
ERROR: test_autoincrement_reset2 (test_runner.tests.AutoIncrementResetTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/django/django/test/testcases.py", line 274, in _setup_and_call
    self._pre_setup()
  File "/django/django/test/testcases.py", line 960, in _pre_setup
    self._fixture_setup()
  File "/django/django/test/testcases.py", line 1002, in _fixture_setup
    self._reset_sequences(db_name)
  File "/django/django/test/testcases.py", line 991, in _reset_sequences 
    no_style(), conn.introspection.sequence_list())
  File "/django/django/db/backends/base/introspection.py", line 130, in sequence_list 
    sequence_list.extend(self.get_sequences(cursor, model._meta.db_table, model._meta.local_fields))
  File "/django/django/db/backends/sqlite3/introspection.py", line 112, in get_sequences 
    pk_col = self.get_primary_key_column(cursor, table_name)
  File "/django/django/db/backends/sqlite3/introspection.py", line 212, in get_primary_key_column
    raise ValueError("Table %s does not exist" % table_name)
ValueError: Table test_runner_person does not exist
```

**Comment by felixxm** [comment]
You can narrow it down to 3 tests, e.g.
```
 ./runtests.py test_runner.tests.AutoIncrementResetTest test_runner.tests.SQLiteInMemoryTestDbs --settings=test_sqlite
```

**Comment by cjerdonek** [comment]
Okay, I pushed a possible fix. The issue is that when running tests in parallel, the test suite was a "mixed" suite, with `ParallelTestSuite` at the root with child suites that were `unittest.TestSuite` objects. So in `iter_test_cases()`, rather than checking the class against the given suite, I switched to check whether the item was an instance of `unittest.TestCase`.

**Comment by cjerdonek** [comment]
(The function at fault was `DiscoverRunner._get_databases()`, which before my change was checking things against `unittest.TestCase` rather than against `type(suite)`. Hence it wasn't susceptible to the "mixed suite" issue.)

**Comment by felixxm** [comment]
@cjerdonek Thanks :+1: I squashed commits and added [small assertions](https://github.com/django/django/pull/14076/commits/22c9af0eae5be38618b5facbb55f5d62c661d9af#diff-921fdc8a429f88233ba9406b266a17d2c90b0e6d205b39f15e2bd78817cf5878R103-R105) which fail without the last change.

**Comment by charettes** [comment]
Thanks for the adjustments folks 👍

**Pull request merged**